torch/_export/serde/type_utils.py

```diff
@@ -0,0 +1,126 @@
+from typing import List, Dict, Tuple, Set, Union
+import inspect
+
+class Succeed:
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, Succeed)
+
+    def __repr__(self) -> str:
+        return "Succeed()"
+
+class Fail:
+    def __init__(self, msg: List[str]) -> None:
+        self.msg = msg
+
+    def __repr__(self) -> str:
+        return f"Fail(msg={self.msg})"
+
+    def __eq__(self, other: object) -> bool:
+        return type(other) == Fail and self.msg == other.msg
+
+
+class NotSure:
+    def __init__(self, v: object, t: object) -> None:
+        self.v = v
+        self.t = t
+
+    def __repr__(self) -> str:
+        return f"NotSure(v={self.v}, t={self.t})"
+
+    def __eq__(self, other: object) -> bool:
+        return type(other) == NotSure and self.v == other.v and self.t == other.t
+
+
+def pp(t: object) -> str:
+    """
+    pretty print a type object
+    """
+
+    head, args = deconstruct(t)
+    if head == list:
+        return f"List[{pp(args[0])}]"
+    elif head == set:
+        return f"Set[{pp(args[0])}]"
+    elif head == dict:
+        return f"Dict[{pp(args[0])}, {pp(args[1])}]"
+    elif head == Union:
+        args_str = ", ".join(pp(arg) for arg in args)
+        return f"Union[{args_str}]"
+    else:
+        return t.__name__
+
+
+def check(v: object, t: object) -> Union[Succeed, Fail, NotSure]:
```

Comment:
Can we add this check to [`serialize`](https://github.com/pytorch/pytorch/blob/cebc11ae8fbc13de4eadee19c579712e5d282754/torch/_export/serde/serialize.py#L1214) and see what breaks?